### PR TITLE
[8.x] Adds once check to `signed` middleware

### PR DESCRIPTION
**Pull request by DarkGhostHunter**
## What?

Allows to process a temporary signed route only once by adding a second parameter to the `signed` middleware: `once`.

```php
$url = URL::temporarySignedRoute('confirm', now()->addMinutes(5), ['email' => 'john@email.com']);

Route::get('email/confirm', function ($request) {
    $request->user()->update('email' => $request->input('email'));

    return 'ok';
})->name('confirm')->middleware('signed:absolute,once');
```

## How?

It adds a relatively simple check to the `ValidateSignature` middleware. If the developer sets `once`, two checks are made: if the `expires` query exists, and if the signature was already handled by checking inside the cache for it.

Once is handled, it will save the signature in the cache, so the next same request will be considered to have an invalid signature.

```php
if ($request->hasValidSignature($relative !== 'relative') && ($once !== 'once' || $this->once($request))) {
    return $next($request);
}

throw new InvalidSignatureException();
```

> This **requires** a signed route to be temporary, since it uses the expiration time to set the TTL of the cache key. If it's not present, the signature won't be considered valid.

The goodies? You can alter the prefix (for the rare case of already using the default for anything else) because it adds an static property. You can also use a Closure in case you need to add like the server name handling the request.

```php
ValidateSignature::$prefix = function($request) {
    return 'signed.once|' . $request->getHost();
};
```

Finally, you can change the cache store to use in the `config.signed` key:

```php
return [
    // ...

    'signed' => 'file',
];
```

## Why?

Because some requests are not idempotent. For example, when an action is cumulative, additional checks must be done in the controller only to stop the request from proceeding. We can avoid adding views or database queries with just four characters: `once`.

Another thing is that, if the URL gets leaked accidentally by the user, once used, it becomes useless for the attacker. This makes theoretically safe for, as example, emailing login links, since the link can't be used again.

## BC?

Extremely low. While this is only cumulative, folks extending the `ValidateSignature` will get a nice error, but... Is there anyone doing that? Is there an use case? Won't be mad if this is pushed to 9.x.

## Shortcomings?

If the very same query is used, the signature won't change, so the user will be blocked. This may be not a very rare case considering a temporary signed route may be set with a static expiration time. It should be noted to not be used in these cases.

```php
// Once the user hits this URL, it will be blocked until next monday:
$url = URL::temporarySignedRoute('invite.accept', Carbon::parse('next monday'), ['user' => 1]);
```

This can be fixed by adding a random string as parameter.

```php
// Adding a random string can help to change the signature:
$url = URL::temporarySignedRoute('invite.accept', Carbon::parse('next monday'), ['user' => 1, 'once' => Str::random()]);
```

### CODE REVIEW

File: src/Illuminate/Routing/Middleware/ValidateSignature.php [hunk 1]
```diff
@@ -2,11 +2,21 @@
 
 namespace Illuminate\Routing\Middleware;
 
+use function cache;
 use Closure;
+use function config;
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Middleware/ValidateSignature.php:7]: can't call a foundation function from the routing package

File: src/Illuminate/Routing/Middleware/ValidateSignature.php [hunk 2]
```diff
@@ -2,11 +2,21 @@
 
 namespace Illuminate\Routing\Middleware;
 
+use function cache;
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Middleware/ValidateSignature.php:5]: same issue

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
I question how useful this is. What if the link is fetched before hand to preview it? Some email clients do this, and does does Android's stock Messages app.

**Comment by DarkGhostHunter** [comment]
> I question how useful this is. What if the link is fetched before hand to preview it? Some email clients do this, and does does Android's stock Messages app.

It seems that iOS does the same. 

An alternative would be making view with a form to a POST signed once route that uses the query. That way the view can be fetched multiple times without hitting the signed once route. 

In that case this PR would belong more to the Foundation package, as it would use the View and Cache.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**